### ISSUE TITLE
Differences between sass-loader v6 and v7

### ISSUE THREAD

**Issue by i-like-robots** [Financial-Times/dotcom-page-kit#269]
In Storybook we have configured v7 of the `sass-loader` package but in our public Sass build plugin we are currently using v6. There are a few key differences between the two which are not straightforward to resolve. This issue is intended to provide an overview of how dependency resolution is performed in Sass and how changes to this algorithm affect our code. It should also be used to track any issues we find and the measures needed to mitigate them.

## Sass partial files

With Sass you can include the contents of partial files using the `@import` directive. If you import another Sass file - usually called "partials" - then the two files will be combined.

The `@import` directive is followed by a path specifier. This path will be resolved relative to the current file. It is a convention to prefix Sass partial files with an underscore `_` but this does not need to be included in the path specifier. The file extension may also be excluded and the Sass implementation will attempt to resolve both `.scss` and `.sass` extensions. This means all of the following path specifiers are equal:

```scss
@import 'path/to/partial';
@import 'path/to/_partial';
@import 'path/to/partial.scss';
@import 'path/to/_partial.scss';
```

This is different to the path specifiers defined by JavaScript module systems. When using Node.js path specifiers must begin with `./` or `../` and absolute paths begin with `/`.

Specifiers which refer to the name of an installed package rather than the path to a file do not need prefixing and are called a "bare specifier". Importing a specific file from a package is called a "deep import".

```js
require('./local/path/specifier')
require('/absolute/path/specifier')
require('bare-specifier')
require('bare-specifier/with/deep/import')
```

## Module resolution

Sass has a simple algorithm to find partial files (whether you're using Ruby Sass, LibSass, or Dart Sass) and you can configure it with a list of directories to fallback to using an option. If Sass cannot resolve a path specifier relative to the current file it will then look in each of the configured directories for it.

Webpack has a far more complicated and powerful [resolution algorithm][7]. It can lookup directories recursively (going up the directory tree from the current context towards the system root), read manifest files with fields defining which files to use, and look for different file names with different file extensions. Each of these settings is also configurable, for example to find packages either installed with Bower or npm you might configure Webpack like this:

```js
{
  ...
  extensions: ['.js', '.mjs']
  resolve: {
    modules: ['bower_components', 'node_modules'],
    descriptionFiles: ['bower.json', 'package.json'],
    mainFields: ['browser', 'module', 'main'],
    mainFiles: ['index', 'main']
  }
}
```

Using this configuration Webpack can generate a long list of potential locations to look for files. 

## `node-sass` vs `sass-loader`

The `sass-loader` package implements a [custom importer][6]. This augments the [built in][4] file resolution capability of `node-sass` (or Dart Sass if you're using that) with the Webpack resolver.

To use the Webpack resolver you can prefix a specifier with a tilde `~`. For example, after installing [o-labels](https://registry.origami.ft.com/components/o-labels@4.2.2?brand=master) it could be referenced like this:

```scss
@import '~o-labels';
@import '~@financial-times/o-labels`;
```

_We have never used this feature._ Origami and Next FT both predate `sass-loader` and did not initially use Webpack. Origami also cannot assume which build system is being used by consumers. This means we depend on the built in partial file resolution of Sass by using a "deep import":

```scss
@import 'o-labels/main';
@import '@financial-times/o-labels/main`;
```

This does not mean `sass-loader` does nothing however, it will still attempt to resolve these files in order to add them to the asset graph and if it is successful it will return the resolved path so that Sass does not have to repeat the effort.

## v6 vs v7

In v6 of `sass-loader` the file extensions used to find partial files are hard-coded. If `sass-loader` is asked to resolve `@import 'main'` it will try to find a file named `main.scss`, `main.sass`, and `main.css` in that order.

Since v7 of `sass-loader` the configuration has [changed][1] and it will use the default list of `extensions` defined in your Webpack configuration. By default Webpack configuration includes JavaScript and JSON extensions only. This can be fixed by appending some extra resolution configuration to the Sass rule:

```diff
{
  rules: [
    {
      test: [/\.s(c|a)ss$/],
+      resolve: {
+        extensions: ['.scss', '.sass', '.css'],
+      },
      use: [
        'style-loader',
        'css-loader',
        'sass-loader'
      ]
    }
  ]
}
```

## Why Bower (and using Origami components) is tricky

Usually we think of Webpack as a tool to bundle JavaScript and it is by default. Out of the box if we try to use Webpack's module resolution algorithm to find Sass files it will fail because they do not use a `.js` extension.

We can add extra configuration to instruct Webpack to look for Sass and CSS files instead but we cannot take full advantage of Webpack's resolver because this algorithm will try to read from manifest files. This won't work for two reasons:

1. Origami components do not specify a field for styles or Sass code in `package.json` (when installed using npm)
2. In `bower.json` it is valid to specify an array of "main" files which Webpack does not understand.  Origami components which provide both styles and JS use this format.

## TL;DR

1. Always use deep import specifiers when referencing Sass partials
2. If you use `sass-loader` v7 you will need to reconfigure the file extensions to resolve for the Sass rule.
3. Hopefully `sass-loader` will stop trying to resolve imports it has not been asked to resolve with `~`

[1]: https://github.com/webpack-contrib/sass-loader/blob/master/CHANGELOG.md#700-2018-04-13
[2]: https://docs.npmjs.com/files/package.json#main
[3]: https://github.com/codewizz/bower-resolve-webpack-plugin
[4]: https://github.com/sass/node-sass#includepaths
[5]: https://www.npmjs.com/package/sass-graph
[6]: https://github.com/sass/node-sass#importer--v200---experimental
[7]: https://webpack.js.org/concepts/module-resolution/
